Add a new module: customer device manager.

loganSIO opened the pull request:
The behavior of this module is designed to empower B2B customers to assign the devices they rent to their employees.

loganSIO committed:
New empty module to help customers manage their devices

loganSIO committed:
Implement device assignment and device history models and related tests

loganSIO committed:
Add security rules, groups, access control, and related tests

loganSIO committed:
Add Device Assignment UI

loganSIO committed:
Refactor update_lot_contract logic

loganSIO committed:
Create Device Assignment lifecycle

Device assignments should not be created nor deleted by customers.
So we create them when a device arrives at their company.
And delete them when the device is returned in our stock.

fcayre commented on customer_device_manager/tests/test_device_assignment.py:
```diff
@@ -0,0 +1,94 @@
+                "partner_id": self.partner1.id,
+            }
+        )
+        locastion_dest = self.env["stock.location"].search([], limit=1)
```
typo à location_dest

fcayre commented on customer_device_manager/tests/test_device_assignment.py:
```diff
@@ -0,0 +1,94 @@
+            }
+        )
+
+    def test_device_assignment_archiving(self):
```
J'ai l'impression que ce test est complètement HS car : 
- le code qui gère l'archivage n'arrive que beaucoup plus tard et il est testé
- sa logique m'échappe de toutes façons
Est-ce que c'est un oubli de suppression ?

loganSIO replied:
Tout à fait d'accord avec toi !
+1 et supprimé*

fcayre commented on customer_device_manager/tests/test_device_assignment.py:
```diff
@@ -0,0 +1,94 @@
+                "product_uom": self.product.uom_id.id,
+                "product_uom_qty": 1,
+                "location_id": locastion_dest.id,
+                "location_dest_id": locastion_dest.id,
```
Les emplacements d'origine et de destination du mouvement sont identiques, normal ? Par ailleurs le mouvement n'est pas une entité habituellement manipulée directement, on utilise en général des pickings, qui créent le mouvement. Il faut revoir la logique de ce test.

fcayre commented on customer_device_manager/models/device_assignment.py:
```diff
@@ -0,0 +1,53 @@
+from odoo import fields, models
+
+
+class DeviceAssignment(models.Model):
```
Ça vaudrait le coup de rajouter une docstring sur la classe pour expliquer brièvement son fonctionnement par rapport à l'historique : le partenaire d'un assignment peut être modifié mais on conserve dans la classe historique la mémoire des changements effectués.

loganSIO replied:
Fait !

fcayre commented on customer_device_manager/tests/test_device_assignment.py:
```diff
@@ -0,0 +1,94 @@
+        self.assertEqual(len(assignment.history_ids), 1)
+
+        assignment.update({"partner_id": self.partner2.id})
+        self.assertTrue(len(assignment.history_ids), 2)
```
Tant que tu y es tu devrais tester ici l'ordre de l'historique en fixant les dates des assignments explicitement et en vérifiant la valeur de assignment.history_ids.mapped("date").

loganSIO replied:
Fait (à re-review)!

fcayre commented on customer_device_manager/security/ir.model.access.csv:
```diff
@@ -0,0 +1,9 @@
+acl_device_assignment_internal,access.device_assignment.internal,model_customer_device_manager_device_assignment,base.group_user,1,0,0,0
+acl_device_assignment_history_internal,access.device_assignment_history.internal,model_customer_device_manager_device_assignment_history,base.group_user,1,0,0,0
+acl_device_assignment_manager,access.device_assignment.manager,model_customer_device_manager_device_assignment,customer_device_manager.group_user_manager,1,1,1,0
+acl_device_assignment_history_manager,access.device_assignment_history.manager,model_customer_device_manager_device_assignment_history,customer_device_manager.group_user_manager,1,0,0,0
```
On fait le choix ici que personne en interne ne peut modifier l'historique, je pense que c'est pas une bonne idée, en cas d'erreur c'est l'IT qui devra corriger... Ajouter tous les droits sur l'historique au group_user_manager me semble nécessaire, qu'en dis-tu ?

loganSIO replied:
Tout à fait d'accord, +1 et fait !

fcayre commented on customer_device_manager/security/rules.xml:
```diff
@@ -0,0 +1,105 @@
+  </record>
+
+  <record model="ir.rule" id="rule_device_assignment_manager">
+    <field name="name">Device Assignment: Managers see all assignments</field>
```
Cette règle est-elle vraiment utile ? Le ir.model.access correspondant devrait faire le job non ?

loganSIO replied:
Exact, règle supprimée.

fcayre commented on customer_device_manager/models/stock_move.py:
```diff
@@ -0,0 +1,36 @@
+    def _set_lot_contract(self, lots, contract, **kwargs):
+        "Create Device Assignment on device expedition validation"
+        super()._set_lot_contract(lots, contract, **kwargs)
+        partner = contract.partner_id.commercial_partner_id
```
ça mérite un petit saut de ligne (de manière générale après un appel à super(), pour qu'il saute aux yeux).
(pas important)

loganSIO replied:
fait sur tous les super() du code. merci!

fcayre commented on customer_device_manager/models/stock_move.py:
```diff
@@ -0,0 +1,36 @@
+                (
+                    "partner_id",
+                    "child_of",
+                    contract.partner_id.commercial_partner_id.id,
```
ajouter une variable `partner = contract.partner_id.commercial_partner_id` pour améliorer la mise en page et faciliter la lecture.
(pas important)

loganSIO replied:
fait, merci!

fcayre commented on customer_device_manager/tests/test_stock_move.py:
```diff
@@ -0,0 +1,132 @@
+        portal_visibility = self.get_visible_assignments(
+            self.lot, self.customer_employee
+        )
+        self.assertFalse(portal_visibility)
```
Est-ce qu'on a besoin de la variable ici ? J'ai l'impression qu'on pourrait se contenter de : 
```
self.assertFalse(self.get_visible_assignments(self.lot, self.customer_employee))
```
(pas important)

loganSIO replied:
fait

fcayre commented on customer_device_manager/tests/test_stock_move.py:
```diff
@@ -0,0 +1,132 @@
+        self.assertFalse(portal_visibility)
+
+        internal_visibility = self.get_visible_assignments(self.lot, self.internal_user)
+        self.assertFalse(internal_visibility)
```
idem
(pas important)

loganSIO replied:
fait

fcayre commented on customer_device_manager/tests/test_stock_move.py:
```diff
@@ -0,0 +1,132 @@
+        portal_visibility = self.get_visible_assignments(
+            self.lot, self.customer_employee
+        )
+        self.assertEqual(len(portal_visibility), 1)
```
idem
(pas important)

loganSIO replied:
fait

fcayre commented on customer_device_manager/tests/test_stock_move.py:
```diff
@@ -0,0 +1,132 @@
+        assigner_visibility = self.get_visible_assignments(
+            self.lot, self.customer_device_assigner
+        )
+        self.assertEqual(len(assigner_visibility), 1)
```
idem
(pas important)

loganSIO replied:
fait

fcayre commented on customer_device_manager/tests/test_stock_move.py:
```diff
@@ -0,0 +1,132 @@
+        with self.assertRaises(AccessError):
+            assignment.sudo(self.customer_employee).update(
+                {
+                    "assignment_reason": "Attempt to update as a customer employee",
```
tu peux compacter un peu le code en mettant tout le dictionnaire sur une ligne
(pas important)

loganSIO replied:
pre-commit me le formatte comme tel

fcayre commented on customer_device_manager/tests/test_stock_move.py:
```diff
@@ -0,0 +1,132 @@
+        portal_visibility = self.get_visible_assignments(
+            self.lot, self.customer_employee
+        )
+        self.assertFalse(portal_visibility)
```
idem
(pas important)

loganSIO replied:
fait

fcayre commented on customer_device_manager/tests/test_stock_move.py:
```diff
@@ -0,0 +1,132 @@
+        assigner_visibility = self.get_visible_assignments(
+            self.lot, self.customer_device_assigner
+        )
+        self.assertFalse(assigner_visibility)
```
idem
(pas important)

loganSIO replied:
fait.

fcayre commented on customer_device_manager/tests/test_stock_move.py:
```diff
@@ -0,0 +1,132 @@
+                [
+                    ("device_id", "=", lot.id),
+                ],
+            )
```
compacter : `.search([("device_id", "=", lot.id)])` me semble bien suffisant
(pas important)

loganSIO replied:
Pre-commit me le format comme tel...

loganSIO reviewed (commented):
Toutes les requests review ont été faites !

loganSIO committed:
Remove active filter from security rules and update record names

loganSIO committed:
Replace active boolean with device location field in device assignments

This commit changes how device tracking is handled by replacing active boolean
with a more descriptive field (device_location) that can track wether a device is
at the customer or at Commown.

loganSIO committed:
Add date_range_portal: Portal access to date ranges

loganSIO committed:
Track device_location on device_assignment history's form

loganSIO committed:
Improve stock_move to update same assignation if necessary instead of creating new one every time

fcayre commented on date_range_portal/tests/test_ir_rules.py:
```diff
@@ -0,0 +1,17 @@
+
+
+class PortalDateRangeTypeIRRulesTC(SavepointCase):
+    def _give_portal_access(self, partner):
```
On peut enlever pas mal de ligne de code en réutilisant un utilisateur du portail existant (genre `env.ref("base.demo_user0")` je crois)

fcayre commented on date_range_portal/tests/test_ir_rules.py:
```diff
@@ -0,0 +1,17 @@
+
+        partner = self.env.ref("base.res_partner_address_15")
+        user = self._give_portal_access(partner)
+        self.env["date.range.type"].sudo(user).search_read([], fields=["name"])
```
Il faudrait s'assurer ici qu'on obtient un record set vide.

fcayre commented on date_range_portal/readme/DESCRIPTION.rst:
```diff
@@ -0,0 +1 @@
+Fix read access on date range type for portal users
```
Je pense qu'il faut compléter en disant qu'on corrige l'erreur en interdisant tout accès mais sans erreur.

fcayre commented on customer_device_manager/models/stock_move.py:
```diff
@@ -11,14 +11,35 @@ def _set_lot_contract(self, lots, contract, **kwargs):
+                            "assignment_date": fields.Datetime.now(),
+                        }
+                    )
+
```
Le code semble OK mais sans test unitaire je ne saurais en être sûr ;-)

fcayre commented on customer_device_manager/models/device_assignment.py:
```diff
@@ -90,6 +90,7 @@ def _inverse_partner_id(self):
                     "assignment_id": rec.id,
                     "date": fields.Datetime.now(),
                     "partner_id": rec.partner_id.id,
+                    "device_location": rec.device_location,
```
Il faudrait ajouter une ligne dans un test unitaire pour s'assurer que l'historique de device_location est bien mis à jour.

loganSIO committed:
Add French and German translation

fcayre commented on customer_device_manager/models/device_assignment.py:
```diff
@@ -0,0 +1,121 @@
+        inverse_name="assignment_id",
+    )
+
+    image_medium = fields.Binary(
```
Le champ serait mieux nommé device_image_medium car on pourrait vouloir ajouter (et afficher) partner_image_medium

fcayre commented on customer_device_manager/models/stock_move.py:
```diff
@@ -0,0 +1,73 @@
+                    [
+                        ("device_id", "=", lot.id),
+                    ],
+                    limit=1,
```
Il manque le tri dans cette requête, c'est important. Il faut un test unitaire d'ailleurs, qui n'est pas présent d'après codecov...

fcayre commented on customer_device_manager/models/stock_move.py:
```diff
@@ -0,0 +1,73 @@
+            )
+
+            if assignment:
+                self.env[
```
Test important manquant d'après codecov

fcayre commented on customer_device_manager/security/rules.xml:
```diff
@@ -0,0 +1,79 @@
+            ('partner_id', 'child_of', user.commercial_partner_id.id),
+        ]
+    </field>
+    <field name="perm_read" eval="True"/>
```
Je pense que cette règle ne devrait concerner que l'écriture, pas la lecture, qui est régie par la règle précédente des utilisateurs du portail (et qui ne contient pas la restriction sur la localisation en passant, ce qui est ce qu'on veut je crois).

loganSIO committed:
Import test files to fix the CI

fcayre commented:
/ocabot rebase

Honeyxilia commented on customer_device_manager/tests/test_stock_move.py:
```diff
@@ -51,51 +52,39 @@ def setUp(self):
-        return self.env["customer_device_manager.device_assignment"].search(
-            [("device_id", "=", lot.id), ("partner_id", "=", company.id)],
+        """Find device assignments for a lot and company, including assignments to contacts of the company."""
+        assignments = self.env["customer_device_manager.device_assignment"].search(
```
est-ce que cette instruction pourrait potentiellement être inlinée ?

Honeyxilia replied:
J'ai vu que c'est fait dans la PR #465 du coup je résout déjà cette convo c:

Honeyxilia commented on customer_device_manager/tests/__init__.py:
```diff
@@ -0,0 +1,3 @@
+from . import test_device_assignment
```
je pense qu'on peut déclarer ce fichier dans les commit où l'on crée les tests

fcayre replied:
Oui carrément, mais on va considérer ça comme secondaire ici et laisser les commits de Logan tels quels du coup vu que c'est la seule remarque : c'est juste plus simple ! Merci pour cette relecture supplémentaire.

Honeyxilia reviewed (approved):
Mis à part des remarques de forme, je n'ai pas plus de remarques c:

fcayre commented:
/ocabot rebase

fcayre commented:
/ocabot merge patch
